[Memory] Use move operator to prevent extra memory allocs

TD-er opened the pull request:
For the controller queue messages, as well as EventStruct (and some more) the generated code was often using a copy constructor instead of the move operator.
This would lead to lots of memory allocations which are not needed.

For the EventStruct and the controller queue element structs there are only move constructors and no copy constructors.
C++ only can generate copy-constructor and move constructor if no user defined move constructor or move assignment operator is present.
So by manually creating one, and not having a copy-constructor and/or assignment operator, you can make sure those copy constructors are not generated and unknowingly used instead of the move operator.

For the EventStruct you still need to copy the struct in a few places, so for those I added a `deep_copy` function so the user must perform an explicit deepcopy and no longer an unintentional one.

There is still a lot to gain, as we still have lots and lots of functions where non-trivial objects are copied instead of moved (e.g. return value of a function)
But for now, the structs with potentially the largest memory blocks are moved instead of copied.

Also fixed the deduplication check on a number of controllers as those were a buggy due to copy/paste without proper checking.

TD-er committed:
[Controllers] Optimize memory usage by using std::move

Controllers now use the move operator where possible to prevent memory allocation when a move is possible.

Also started implementation of checking for message count per unit to see if we received the exact message twice.

TD-er committed:
[Controller] Do not keep unitMessageCount as member when not needed

TD-er committed:
Add move assignment to EventStruct

TD-er committed:
[Events] Make sure events are not copied unless absolutely needed

TD-er committed:
[Controllers] Implement move constructor to guarantee no copy is made

TD-er committed:
[MQTT] Fix MQTT controller may not connect when re-enabled after a while

TD-er committed:
[Memory] getMaxFreeBlock() should return valid value on ESP32

mcspr commented on src/_C001.ino:
```diff
@@ -111,7 +111,7 @@ bool CPlugin_001(CPlugin::Function function, struct EventStruct *event, String&
             # endif // if FEATURE_ADC_VCC
 
-          success = C001_DelayHandler->addToQueue(C001_queue_element(event->ControllerIndex, event->TaskIndex, url));
+          success = C001_DelayHandler->addToQueue(std::move(C001_queue_element(event->ControllerIndex, event->TaskIndex, std::move(url))));
```
std::move() for in-place object ctor (aka it will be destroyed in the current block when on the next line) or function call returning object could be safely omitted btw, it should already be && as presented by the compiler

TD-er replied:
Yep, is probably a left-over as I did this first, then realized it still might use the copy constructor and then I added a move constructor to make sure copy constructor is never used.
I will check if I can remove this `std::move` wrapper and see how the compiler thinks of it.

mcspr replied:
cppreference also describes these as prvalues:
https://en.cppreference.com/w/cpp/language/value_category#C.2B.2B11
aka a temporary, which I forgot a word for

also tangentially related, rvalue stops being rvalue in the method / function block itself:
https://godbolt.org/z/3EPcPhePf
so copy might accidentally happen after accepting && and calling something else expecting &&, but not using `std::move(value)` / `static_cast<T&&>(value)`

TD-er replied:
Yep, so that's why the copy constructor or assignment operator should be declared with `= delete` where possible to prevent these to make sure the compiler will catch them.
Still tricky on objects that have all declared like `String`.

But as you may have seen, I have added a lot of those compiler checks in the code to help finding bugs (and now performance hits) as soon as possible.
So I'm happy to have learned about the `=delete` and `=default`, which may help a lot.

TD-er committed:
[Network] Properly switch between ETH and WiFi

ESP32 does send events for connected/disconnected only once.
So we cannot rely on the events.

Also the WiFi disconnect was not handled well when switching between network media.

TD-er committed:
[WiFi] Fix small issues showing WiFi state & abort WiFi connection process

TD-er committed:
[WiFi] Show WiFi STA MAC in sysinfo when eth is active

TD-er committed:
[Network] Allow fallback to WiFi when Ethernet not connected at boot

TD-er commented:
Also added quite a lot of fixes regarding WiFi/Ethernet switching and handling unplanned WiFi disconnects.

mcspr commented on src/src/ControllerQueue/C011_queue_element.cpp:
```diff
@@ -6,6 +6,18 @@
 
 C011_queue_element::C011_queue_element() {}
 
+C011_queue_element::C011_queue_element(C011_queue_element&& other)
+  : uri(std::move(other.uri))
+  , HttpMethod(std::move(other.HttpMethod))
+  , header(std::move(other.header))
+  , postStr(std::move(other.postStr))
+  , idx(other.idx)
+  , _timestamp(other._timestamp)
+  , TaskIndex(other.TaskIndex)
+  , controller_idx(other.controller_idx)
+  , sensorType(other.sensorType)
+{}
+
```
In general, *one should* prefer `=default` at least when ctor only deals with member copy and does not do anything in the body:
- no need to update ctor each time members are changed
- no way to forget members have changed
- noexcept for manually-typed ctros needs to be here to avoid copies with std containers which check for `std::is_nothrow_move_constructible<ContainedClass>`-> https://en.cppreference.com/w/cpp/types/is_move_constructible, but idk if this applies here)

TD-er replied:
Sounds like a good idea, will have a look at it.
To be honest, the move operator itself is still quite new to me, so glad there is something like the =default (as it does like I think it does)

mcspr commented on src/src/ControllerQueue/C015_queue_element.cpp:
```diff
@@ -6,6 +6,17 @@
+  , controller_idx(other.controller_idx), valuesSent(other.valuesSent)
+  , valueCount(other.valueCount)
+{
+  for (byte i = 0; i < VARS_PER_TASK; ++i) {
+    txt[i]  = std::move(other.txt[i]);
+    vPin[i] = other.vPin[i];
+  }
```
ref. above, this also can be omitted in case `txt = std::move(other.txt);` could work (or copy), consider `std::array<String, VARS_PER_TASK>;`? or a custom object implementing `Object& operator=(Object&&) noexcept;'

mcspr reviewed (commented):
iirc, some minor notes regarding move usage itself, as I have no comments regarding the features added / changed

mcspr commented:
> So by manually creating one, and not having a copy-constructor and/or assignment operator, you can make sure those copy constructors are not generated and unknowingly used instead of the move operator.

Also note of the `=delete;` for everything that allows `=default;`, which will trigger compilation errors when it is used by some code

TD-er commented:
> > So by manually creating one, and not having a copy-constructor and/or assignment operator, you can make sure those copy constructors are not generated and unknowingly used instead of the move operator.
> 
> Also note of the `=delete;` for everything that allows `=default;`, which will trigger compilation errors when it is used by some code

Sounds also like a good concept to read more about :)

TD-er committed:
[Oopsie] Need to hide in other commits, deny this ever was committed!

As reported on [the forum](https://www.letscontrolit.com/forum/viewtopic.php?f=6&t=8528&p=52439#p52439)

TD-er committed:
[Cleanup] Remove unneeded std::move and copy/assignment constructors

TD-er committed:
[Controller Queue] Remove unneeded constructors using =default and =delete

TD-er commented:
Applied it to EventStruct and it makes it a lot more clean :)
I have to think about the arrays of members, like you suggested, but will for sure also cleanup those later.
